### ISSUE TITLE
Controlling where TBB threads are running

### ISSUE THREAD

**Issue by dimstamat** [01org/tbb#88]
I have been looking for a way to assign TBB threads to specific hardware cores, but after some research I saw that TBB does not give any control to where the TBB threads are running.
My current project has the requirement of assigning the TBB threads to specific CPU cores and as I found at software intel forums and online, many more people have the same requirement.
There is a way to do it with thread local storage:
https://software.intel.com/en-us/forums/intel-threading-building-blocks/topic/294546
But this requires each task to check its thread local storage and determine whether the current thread is pinned to a hardware core, making it costly.
It also requires some sort of synchronization among threads so that to assign the right core IDs.
My solution is doing the thread pinning at thread creation within src/rml/serve/thread_monitor.h. I am using compare-and-swap instruction to access a shared CPU number array and make each thread get one of these numbers.
May I open a pull request to present my solution?
Please let me know what you think.
Thanks,
Dimos

**Comment by anton-malakhov** [comment]
That forum thread is very old. Have noticed "local observer" feature yet? https://www.threadingbuildingblocks.org/docs/help/reference/appendices/preview_features/task_scheduler_observer_ext_cls.html
It basically enables you to pin threads according to their task affinity id which enables affinity_partitioner and pure affinitized tbb::task to choose on which hardware to run. It has much less overhead than checking affinity in each task but still not quite optimal since worker threads tend to leave arena and can re-enter on a different slot when coming back for a work, so between execution of parallel algorithms, it causes additional latency for threads to join the work. This can be partially solved using `may_sleep` callback which is part of the same local observer interface but requires preview tbb binary (if it helps you a lot - please provide feedback so TBB team can move this feature to production binary).

FYI, an example of an elaborated affinity&pinning-enabled algorithm was reported here: https://software.intel.com/sites/default/files/managed/9d/48/ConfAnton-Pushing-the-limits-of-work-stealing-approved.pdf

**Comment by dimstamat** [comment]
Thanks for your response! That thread is really old indeed. I would be surprised if nothing was available to support thread affinity in TBB as of today! :)
Too bad I didn't see this thread on time, but now I already have my own solution.
However, by overriding the task_scheduler_observer:: on_scheduler_entry you call set affinity on every newly started task, right? Isn't this similar with the example in the old thread?
My solution sets the thread affinity only once, at startup. This could also be a parameter at task_scheduler_init.
Thanks!

**Comment by alexey-katranov** [comment]
Consider the blog article about [how to pin threads with TBB](https://software.intel.com/en-us/blogs/2013/10/31/applying-intel-threading-building-blocks-observers-for-thread-affinity-on-intel). This article demonstrates the "local observer" idea on Intel Xeon Phi but it can be apllied easily for almost any Linux. If you need Windows based solution you need to update affinity related system API.

> May I open a pull request to present my solution?

If you are asking a permission, you do not need it. Feel free to open pull request for any reasonable idea (pay attention that it can be rejected). If you are asking that the idea will be accepted, at first glance, it does not look the right approach due to several reasons:

1. Correctness: TBB can create more threads that the number of CPUs in the system (even if you create the `task_scheduler_init` object). Does the approach survive this occasion?
2. Composabilty: is the proposed affinity approach usefull for all applications? Will it be enabled by default or by request? What if one component wants affinity but other not?
3. Optimization/Complexity: the "compare-and-swap instruction to access a shared CPU number array" does not seems the best approach. Each thread has [its own index](https://github.com/01org/tbb/blob/tbb_2019/src/tbb/private_server.cpp#L67) that can be used to choose the right CPU.

Therefore, TBB introduced the local_observer mechanism to overcome these issues.

**Comment by dimstamat** [comment]
Thanks alexey!
1. Yes, my approach makes sure that the number of requested threads is less than the number of CPUs in the system, or else it does not apply the thread affinity.
2. It will be enabled by request. That could be an issue as of which component wants affinity and which doesn't.
3. You have to use some sort of synchronization there because the [launch function](https://github.com/01org/tbb/blob/tbb_2019/src/rml/server/thread_monitor.h#L212) is called by multiple threads. By just using the my_index will result to race conditions and incorrect results. (Multiple threads will be trying to set the same ID). Also, std::atomic is not sufficient because someone might get the same ID before the first thread increments it, leading to 2 threads setting the same ID.

Anyhow, I just wanted to point this out as there was lots of demand for thread affinity after a google search.
It seems like your provided solution does the job!
Thanks a lot for your replies!

**Comment by dimstamat** [comment]
OK then, that's great! But, does "when a thread is ready to work" mean when a task finishes and the thread wants to start a new task?

**Comment by anton-malakhov** [comment]
No. When a task arena has no more tasks for any thread, it will be marked empty and threads will start to fall asleep, leaving task arena. When a new task is created for an empty arena, threads will receive wake-up signal and come back looking for a non-empty task arena. The local observer notification is received when threads are joining and leaving task arena.

BTW, Alexey's blog is about global observer, not local one. the difference for you is whether you want to recommend where specifically to execute a task (possible only with local observer) or generally want to just pin TBB worker threads (global observer like in the blog is enough).

**Comment by dimstamat** [comment]
Okay then, that sounds perfect! I kinda got the idea of arena, global and local observer when tracing the code to see when new threads are created. For me I think global observer should be enough. I just wanted to pin TBB threads to specific cores. I think the issue is resolved! Thanks again guys!

**Comment by dimstamat** [comment]
A question for Alexey's blog:
is `tbb::task_scheduler_observer` initialized once per TBB application?
Thanks!

**Comment by alexey-katranov** [comment]
You need to create a `tbb::task_scheduler_observer` object and remain it active while you want to receive the notifications from TBB. You can create as many observer objects as you want.
In the blog, "global" observer is used. I'd recommend to replace it with ["local" observer](https://www.threadingbuildingblocks.org/docs/help/reference/appendices/preview_features/task_scheduler_observer_ext_cls.html) (pass `true` to the base class in the constructor).

**Comment by dimstamat** [comment]
Got it! So this means that `on_scheduler_entry` can be called several times concurrently (for each worker thread)? From your doc I see:
"`tbb::internal::task_scheduler_observer_v3::on_scheduler_entry ( bool )`:Invoked from inside observe(true) call and whenever a worker enters the arena this observer is associated with".
I see you are using a `tbb::atomic` variable for `thread_index` accessed by concurrent threads. Is the `tbb::atomic` any different than the `std::atomic`? If not, then how can the following work properly?
`int thr_idx = thread_index++;`
This statement is not entirely atomic. You first do the read of `thread_index`'s value and there might be a context switch, making another thread to also read the same value before you increment. Then, both threads will end up with the same thread index! (Maybe consider `std::atomic::fetch_add()` ?)
I am guessing you mostly use the TBB's current slot (`USE_TASK_ARENA_CURRENT_SLOT` is set), or maybe `tbb:atomic` is different than `std::atomic`?
Thanks for your comments!

**Comment by alexey-katranov** [comment]
`tbb::internal::task_scheduler_observer_v3::on_scheduler_entry` is really called concurrently for worker and master threads. Moreover, it can be called more than one time for each particular thread (paired with `tbb::internal::task_scheduler_observer_v3::on_scheduler_exit` )

As for `std/tbb::atomic`, they are similar in many aspects. The [increment operation](https://en.cppreference.com/w/cpp/atomic/atomic/operator_arith) is simplified syntax for the [fetch_add](https://en.cppreference.com/w/cpp/atomic/atomic/fetch_add) operation that guarantees atomicity (see [read-modify-write](https://en.wikipedia.org/wiki/Read-modify-write)).

**Comment by zyrch** [comment]
How does one create a global observer? The blog link mentioned is not working and the [documentation](https://spec.oneapi.com/versions/latest/elements/oneTBB/source/task_scheduler/task_arena/task_scheduler_observer_cls.html)  says that the default constructor creates a observer to the owner thread's arena. On using arguments true or false gives a no matching constructor error. Please share if there is something that I am missing here.

**Comment by alexey-katranov** [comment]
> On using arguments true or false gives a no matching constructor error.

@zyrch, oneTBB does not support global observers, so there is no such argument (this thread is related to TBB).

Can you please describe the problem that you trying to solve?

**Comment by zyrch** [comment]
It's the same as the one mentioned in this issue, binding threads to cores. However, since in an arena there is a max concurrency limit and also threads can join and leave multiple times, I was thinking a global observer would be much more efficient.

**Comment by alexey-katranov** [comment]
You can add a flag into `thread_local` that a thread is already bound to the mask. However, pay attention to composability issues if you do not control the whole application and your library can be used with other libraries that also bind the threads.

**Comment by zyrch** [comment]
Thanks. Can you share how does this approach handle the concurrency limit of the arena? I know that I can change the limit to any number I like while creating an arena, but I want this method to be dynamic.

**Comment by alexey-katranov** [comment]
There is no possibility to change the limit for already created `task_arena` (beside destroying and creating with a new limit). However, you can set relatively big limit for the `task_arena` and apply `global_control` to limit the active number of worker threads. `global_control` can be used in dynamic.

**Comment by zyrch** [comment]
Does oneTBB have methods defined to set affinity masks for threads? Something like mentioned [here](https://spec.oneapi.com/versions/latest/elements/oneTBB/source/task_scheduler/task_arena/task_scheduler_observer_cls.html) in the example at the bottom. Does ```src/tbbbind/tbb_bind.cpp``` has something to do with this?
